test/e2e/storage/drivers/in_tree.go

```diff
@@ -178,29 +164,39 @@ func (n *nfsDriver) CreateVolume(config *testsuites.PerTestConfig, volType testp
 	cs := f.ClientSet
 	ns := f.Namespace
 
-	// NewNFSServer creates a pod for InlineVolume and PreprovisionedPV,
-	// and startExternalProvisioner creates a pods for DynamicPV.
-	// Therefore, we need a different PrepareTest logic for volType.
-	switch volType {
-	case testpatterns.InlineVolume:
-		fallthrough
-	case testpatterns.PreprovisionedPV:
-		c, serverPod, serverIP := framework.NewNFSServer(cs, ns.Name, []string{})
-		config.ServerConfig = &c
-		return &nfsVolume{
-			serverIP:  serverIP,
-			serverPod: serverPod,
-			f:         f,
-		}
-	case testpatterns.DynamicPV:
-		// Do nothing
-	default:
-		framework.Failf("Unsupported volType:%v is specified", volType)
+	c, serverPod, serverIP := framework.NewNFSServer(cs, ns.Name, []string{})
+	config.ServerConfig = &c
+
+	// Create an nfs-client-provisioner pod to dynamically mkdir volumes in the server pod
+	var externalProvisionerPod *v1.Pod
+	if volType == testpatterns.DynamicPV {
```

Comment:
Thought more about it, CreateVolume is a method for Preprovisioned drivers, calling it with volType==DynamicPV is probably the wrong thing to do. I would be better to always start the provisioner in PrepareTest even if it's not needed, that's basically what the CSI drivers do. Why should TestDriver.CreateVolume accept a volType argument at all? cc @pohly

Reply:
> Thought more about it, CreateVolume is a method for Preprovisioned drivers, calling it with volType==DynamicPV is probably the wrong thing to do. 

It's a bit confusing, but each driver's ```CreateVolume``` is called by driveroperations.go's ```CreateVolume```.
And driveroperations.go's one calls driver's one, only when it is not DynamicProvision case.
So, it is doing right way, although we might make it easier to understand, by renaming, etc ...


> Why should TestDriver.CreateVolume accept a volType argument at all?

Good point. We would be able to remove this argument, for there won't be different action needed in TestDriver.CreateVolume depending on volType.

See:
  https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/base.go#L174
  https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/driveroperations.go#L46